README
NAME
    MooseX::PrivateSetters - Names your accessors foo() and _set_foo()

SYNOPSIS
        use Moose;
        use MooseX::PrivateSetters;

        # make some attributes

        has 'foo' => ( is => 'rw' );
        ...

        package main;

        sub bar {
            $self = shift;
            $self->_set_foo(1) unless $self->foo
        }

DESCRIPTION
    This module does not provide any methods. Simply loading it changes the
    default naming policy for the loading class so that accessors are
    separated into get and set methods. The get methods have the same name
    as the accessor, while set methods are prefixed with "_set_".

    If you declare an attribute with a leading underscore, then the
    set method will start with "_set" (i.e. C<<has '_foo'>> will get a
    setter called '_set_foo'.

    If you explicitly set a "reader" or "writer" name when creating an
    attribute, then that attribute's naming scheme is left unchanged.

    Load order of this module is important. It must be "use"d after Moose.

SEE ALSO
    There are a number of similar modules on the CPAN.

  Moose
    The original. A single mutator method is generated with the same name as
    the attribute itself.

  MooseX::Accessors::ReadWritePrivate
    Changes the parsing of the "is" clause, making new options available.
    For example,

        has baz => ( is => 'rpwp', # private reader, private writer

    gets you "_get_baz()" and "_set_baz".

  MooseX::FollowPBP
    Names accessors in the style recommended by *Perl Best Practices*:
    "get_size" and "set_size".

  MooseX::SemiAffordanceAccessor
    Has separate methods for getting and setting. The getter has the same
    name as the attribute, and the setter is prefixed with "set_".

AUTHOR
    brian greenfield "<briang@cpan.org>"

BUGS
    Please report any bugs or feature requests to
    "bug-moosex-privatesetters@rt.cpan.org", or through the web interface at
    <http://rt.cpan.org>. I will be notified, and then you'll automatically
    be notified of progress on your bug as I make changes.

COPYRIGHT & LICENSE
    Copyright 2010 brian greenfield

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



CHANGES
0.07    2011-08-28
        - Fix silly spelling of declare. Reported by confound via IRC.
        - Clarified prefix used when attribute is declared with a
          leading underscore. Reported by confound via IRC.

0.06    2011-08-28
        - Fix silly spelling of SemiAffordanceAccessors
          (RT#70560). Reported by Shawn H. Corey. Thanks!

0.05    2011-05-09
        - Fix minimum perl version (RT#67749). Now requires perl 5.8.3

0.04    2011-04-04
        - Make compatable with upcoming Moose 2.0 (code provided my
          Dave Rolsky). See
          http://blog.moose.perl.org/2011/03/changes-to-moosex-modules-with-moose-20.html
          for details.

0.03    2010-07-25
	- Minor documentation changes

0.02    2010-07-24
        - Update TODO
        - Improve documentation
        - Modernise by using class_metaroles in
          Moose::Exporter->setup_import_methods
        - Add tests for attributes starting with double underscore
        - Bump Moose prereq to 0.94, add repository
        - add TODO.org

0.01    2010-07-15
        - Initial release
        - Based on Dave Rolsky's MooseX-SemiAffordanceAccessor-0.05


Build.PL
use strict;
use warnings;

require 5.008003;

use Module::Build;

my $builder = Module::Build->new(
    build_requires     => { 'Test::More'    => '0', },
    configure_requires => { 'Module::Build' => '0', },
    create_makefile_pl => 'traditional',
    create_readme      => 1,
    license            => 'perl',
    meta_merge         => {
        resources      => {
            repository => 'http://github.com/briang/MooseX-PrivateSetters'
        }
    },
    module_name        => 'MooseX::PrivateSetters',
    recommends         => {
        'Config'            => 0,
        'FindBin'           => 0,
        'Module::CoreList'  => 0,
        'Module::Info'      => 0,
        'Parse::CPAN::Meta' => 0,
    },
    requires           => {
        'perl'                  => '5.8.3',
        'Moose'                 => 0.94,
        'Moose::Exporter'       => 0.94,
        'Moose::Role'           => 0.94,
        'Moose::Util::MetaRole' => 0.94,
    },
);

$builder->create_build_script();


lib/MooseX/PrivateSetters/Role/Attribute.pm
package MooseX::PrivateSetters::Role::Attribute;

use strict;
use warnings;

our $VERSION = '0.07';

use Moose::Role;

before '_process_options' => sub {
    my $class   = shift;
    my $name    = shift;
    my $options = shift;

    if ( exists $options->{is} &&
           ! ( exists $options->{reader} || exists $options->{writer} ) ) {
        if ( $options->{is} eq 'ro' ) {
            $options->{reader} = $name;
            delete $options->{is};
        }
        elsif ( $options->{is} eq 'rw' ) {
            $options->{reader} = $name;
            my $prefix = $name =~ /^_/  ?  '_set'  :  '_set_';
            $options->{writer} = $prefix . $name;

            delete $options->{is};
        }
    }
};

no Moose::Role;

1;

=head1 NAME

MooseX::PrivateSetters::Role::Attribute - Names setters as such, and makes them private

=head1 SYNOPSIS

    Moose::Exporter->setup_import_methods(
        class_metaroles => {
            attribute => ['MooseX::PrivateSetters::Role::Attribute'],
        },
    );

=head1 DESCRIPTION

This role applies a method modifier to the C<_process_options()>
method, and tweaks the writer parameters so that they are private with
an explicit '_set_attr' method. Getters are left unchanged. This role
copes with attributes intended to be private (ie, starts with an
underscore), with no double-underscore in the setter.

For example:

    | Code                      | Reader | Writer      |
    |---------------------------+--------+-------------|
    | has 'baz'  => (is 'rw');  | baz()  | _set_baz()  |
    | has 'baz'  => (is 'ro');  | baz()  |             |
    | has '_baz' => (is 'rw');  | _baz() | _set_baz()  |
    | has '__baz' => (is 'rw'); | _baz() | _set__baz() |

You probably don't want to use this module. You probably should be
looking at L<MooseX::PrivateSetters> instead.

=head1 AUTHOR

brian greenfield C<< <briang@cpan.org> >>

=head1 COPYRIGHT & LICENSE

Copyright 2010 brian greenfield

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


lib/MooseX/PrivateSetters.pm
package MooseX::PrivateSetters;

use strict;
use warnings;

our $VERSION = '0.07';

use Moose 0.94 ();
use Moose::Exporter;
use Moose::Util::MetaRole;
use MooseX::PrivateSetters::Role::Attribute;

my %metaroles = (
    class_metaroles => {
        attribute => ['MooseX::PrivateSetters::Role::Attribute'],
    },
);

$metaroles{role_metaroles} = {
    applied_attribute => ['MooseX::PrivateSetters::Role::Attribute'],
} if $Moose::VERSION >= 1.9900;

Moose::Exporter->setup_import_methods(%metaroles);

1;

__END__

=pod

=head1 NAME

MooseX::PrivateSetters - Names your accessors foo() and _set_foo()

=head1 SYNOPSIS

    use Moose;
    use MooseX::PrivateSetters;

    # make some attributes

    has 'foo' => ( is => 'rw' );
    ...

    package main;

    sub bar {
        $self = shift;
        $self->_set_foo(1) unless $self->foo
    }

=head1 DESCRIPTION

This module does not provide any methods. Simply loading it changes
the default naming policy for the loading class so that accessors are
separated into get and set methods. The get methods have the same name
as the accessor, while set methods are prefixed with C<_set_>.

If you declare an attribute with a leading underscore, then the set
method will start with "_set" (i.e. C<has '_foo'> will get a setter
called C<_set_foo>. Note that if you have attributes C<foo> and
C<_foo>, they will both get a setter called C<_set_foo>, so don't do
that.

If you explicitly set a C<reader> or C<writer> name when creating
an attribute, then that attribute's naming scheme is left unchanged.

Load order of this module is important. It must be C<use>d after
L<Moose>.

=head1 SEE ALSO

There are a number of similar modules on the CPAN.

=head2 L<Moose>

The original. A single mutator method is generated with the same name
as the attribute itself.

=head2 L<MooseX::Accessors::ReadWritePrivate>

Changes the parsing of the C<is> clause, making new options
available. For example,

    has baz => ( is => 'rpwp', # private reader, private writer

gets you C<_get_baz()> and C<_set_baz>.

=head2 L<MooseX::FollowPBP>

Names accessors in the style recommended by I<Perl Best Practices>:
C<get_size> and C<set_size>.

=head2 L<MooseX::SemiAffordanceAccessor>

Has separate methods for getting and setting. The getter has the same
name as the attribute, and the setter is prefixed with C<set_>.

=head1 AUTHOR

brian greenfield C<< <briang@cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-moosex-privatesetters@rt.cpan.org>, or through the web interface
at L<http://rt.cpan.org>. I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 COPYRIGHT & LICENSE

Copyright 2010 brian greenfield

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
